Iceberg REST client reorganization

mmaslankaprv opened the pull request:
This is the first PR in the series introducing implementation of Iceberg REST catalog. This PR includes mostly a small reorg of REST client type hierarchy and dependencies. This work is done not to leak the implementation details to the caller and make it easier to extend the client in future. 

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

rockwotj commented on src/v/iceberg/rest_client/oauth_token.h:
```diff
@@ -9,7 +9,10 @@ namespace iceberg::rest_client {
 // Oauth token returned by the catalog server, in exchange for credentials
 struct oauth_token {
     ss::sstring token;
+    ss::sstring token_type;
```
This feels like there is some enum here - can we add a comment about what kind of values are expected here?

rockwotj commented on src/v/iceberg/rest_client/json.cc:
```diff
@@ -0,0 +1,32 @@
+namespace iceberg::rest_client {
+oauth_token parse_oauth_token(const json::Document& doc) {
+    oauth_token ret;
+    ret.token = parse_required_str(doc, "access_token");
```
Should we rename `token` to `access_token` to make it a little more clear?

mmaslankaprv commented:
> the commit
> 
> > iceberg/rc: added missing fields to oauth_token
> 
> doesn't compile:
> 
> ```
> In file included from src/v/iceberg/rest_client/parsers.cc:11:
> bazel-out/k8-fastbuild/bin/src/v/iceberg/rest_client/_virtual_includes/rest_client_parsers/iceberg/rest_client/parsers.h:13:10: fatal error: 'iceberg/rest_client/oauth_token.h' file not found
>    13 | #include "iceberg/rest_client/oauth_token.h"
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 1 error generated.
> ```

fixed

andrwng commented on src/v/iceberg/rest_client/oauth_token.h:
```diff
@@ -0,0 +1,15 @@
+#pragma once
```
nit: license header

andrwng commented on src/v/iceberg/rest_client/oauth_token.h:
```diff
@@ -9,7 +9,10 @@ namespace iceberg::rest_client {
 // Oauth token returned by the catalog server, in exchange for credentials
 struct oauth_token {
```
I don't think we need to change this just yet since presumably Polaris already supports it, but heads up that it looks like this part of the open API spec has been deprecated in favor of pinging the oauth2 server directly instead of the REST catalog. Others in the ecosystem are starting to move as well.

e.g. trino/issues/ 23086

mmaslankaprv replied:
yes exactly, this is being deprecated, it seems this is still going to be useful for testing

bharathv commented on src/v/iceberg/rest_client/catalog_client.h:
```diff
@@ -118,7 +108,7 @@ class catalog_client {
       std::optional<iobuf> payload = std::nullopt);
 
-    std::reference_wrapper<client_source> _client_source;
+    std::unique_ptr<http::abstract_client> _http_client;
```
shouldn't we still call _http_client->shutdown_and_stop() on cleanup or am missing something?

andrwng commented on src/v/iceberg/rest_client/tests/parser_tests.cc:
```diff
@@ -1,56 +0,0 @@
-    ASSERT_FALSE(result.has_value());
-    ASSERT_THAT(
-      result.error(),
-      VariantWith<r::json_parse_error>(Field(
-        &r::json_parse_error::context,
-        "parse_oauth_token::response_does_not_match_schema")));
-}
```
I might be missing them but were these moved somewhere? or are they removed

mmaslankaprv replied:
this particular ones were removed, let me re-write them with the new APIs

andrwng commented on src/v/iceberg/rest_client/catalog_client.cc:
```diff
@@ -123,11 +123,9 @@ ss::future<expected<iobuf>> catalog_client::perform_request(
     while (true) {
         const auto permit = rtc.retry();
         if (!permit.is_allowed) {
-            co_await client_ptr->shutdown_and_stop();
```
Curious where these are called now? I suppose the catalog client needs its own stop method now that it's full owner of the HTTP client

mmaslankaprv replied:
i have a commit introducing shutdown method, in the client public interface, missed adding it to this PR
